apps/build-onchain-apps/src/contract/ContractSpecification.ts

```diff
@@ -1,38 +1,69 @@
 import { Abi, type Chain } from 'viem';
 import { baseGoerli } from 'viem/chains';
+import { useNetwork } from 'wagmi';
 import BuyMeACoffeeABI from './BuyMeACoffee';
 import Custom1155ABI from './Custom1155';
 
-export type ContractInstance = {
+type ContractInstance = {
   chain: Chain;
   address: `0x${string}`;
   deactivated?: boolean;
 };
 
-export type ContractSpecification = Record<
-  string,
-  {
-    abi: Abi;
-    [chainId: number]: ContractInstance;
+type UseContractReturn<T extends Abi> =
+  | { abi: T; address: `0x${string}`; status: 'ready' }
+  | { abi: T; status: 'onUnsupportedNetwork' }
+  | { abi: T; status: 'notConnected' }
+  | { abi: T; status: 'deactivated' };
+
+type Spec<T extends Abi> = {
+  abi: T;
+  [chainId: number]: ContractInstance;
+};
+
+export function generateContractHook<T extends Abi>({ abi, ...spec }: Spec<T>) {
```

Comment:
generating a hook per contract makes things a little more succinct at the callsite, while not being significantly more verbose than the previous plan of single big object of all contracts with a hook that selects from them.

Reply:
move this hook under `onchain/hooks` folder, as this can be an interesting reusable hook.